File: ui/src/androidTest/java/com/xmartlabs/bigbang/ui/recyclerview/singleitem/SimpleItemTestAddItems.java
package com.xmartlabs.bigbang.ui.recyclerview.singleitem;

import com.annimon.stream.Stream;
import com.xmartlabs.bigbang.ui.recyclerview.common.Car;

import org.junit.Test;

import java.util.Arrays;
import java.util.Collections;
import java.util.List;

/**
 * Created by mirland on 28/04/17.
 */
public class SimpleItemTestAddItems extends SimpleItemRecyclerViewTest {
  private void testAddItem(List<Car> initialItems, List<Car> itemsToAdd) {
    SingleItemActivity activity = mActivityRule.getActivity();
    CarAdapter activityCarAdapter = activity.getAdapter();
    activity.runOnUiThread(() -> activityCarAdapter.setItems(initialItems));

    checkItems(initialItems);
    activity.runOnUiThread(() -> activityCarAdapter.addItems(itemsToAdd));

    List<Car> resultList = Stream
        .concat(
            Stream.of(initialItems),
            Stream.of(itemsToAdd))
        .toList();
    checkItems(resultList);
  }

  @Test
  public void testAddOneItem() {
    testAddItem(getCarList(), Collections.singletonList(new Car("New car")));
  }

  @Test
  public void testAddOneItemToAnEmptyList() {
    testAddItem(Collections.emptyList(), Collections.singletonList(new Car("New car")));
  }

  @Test
  public void testAddMultipleItem() {
    List<Car> cars = getCarList();
    Car car1 = new Car("New Car");
    Car car2 = new Car("New Car2");
    List<Car> carList = Arrays.asList(car1, car2);
    testAddItem(cars, carList);
  }

  @Test
  public void testAddOneItemTwice() {
    List<Car> cars = getCarList();
    Car car1 = new Car("New Car");
    List<Car> carList = Arrays.asList(car1, car1);
    testAddItem(cars, carList);
  }
}

File: ui/src/androidTest/java/com/xmartlabs/bigbang/mvp/recyclerview/singleitem/SimpleItemTestAddItems.kt
package com.xmartlabs.bigbang.mvp.recyclerview.singleitem

import com.xmartlabs.bigbang.mvp.recyclerview.common.Car
import org.junit.Test
import java.util.Arrays

class SimpleItemTestAddItems : SimpleItemRecyclerViewTest() {
  private fun testAddItem(initialItems: List<Car>, itemsToAdd: List<Car>) {
    val activity = activityRule.activity
    val activityCarAdapter = activity.adapter
    activity.runOnUiThread { activityCarAdapter.setItems(initialItems) }

    checkItems(initialItems)
    activity.runOnUiThread { activityCarAdapter.addItems(itemsToAdd) }

    val resultList = initialItems.toMutableList().apply { addAll(itemsToAdd) }
    checkItems(resultList)
  }

  @Test
  fun testAddOneItem() {
    testAddItem(carList, listOf(Car("New car")))
  }

  @Test
  fun testAddOneItemToAnEmptyList() {
    testAddItem(emptyList(), listOf(Car("New car")))
  }

  @Test
  fun testAddMultipleItem() {
    val cars = carList
    val car1 = Car("New Car")
    val car2 = Car("New Car2")
    val carList = Arrays.asList(car1, car2)
    testAddItem(cars, carList)
  }

  @Test
  fun testAddOneItemTwice() {
    val cars = carList
    val car1 = Car("New Car")
    val carList = Arrays.asList(car1, car1)
    testAddItem(cars, carList)
  }
}
